helix-loader/src/grammar.rs

```diff
@@ -85,13 +86,15 @@ pub fn get_language(name: &str) -> Result<Language> {
     Ok(language)
 }
 
+#[cfg(not(target_arch = "wasm32"))]
```

Comment:
All of this git/build handling seems like it could be split into a separate file/module then gated behind a single conditional

Reply:
I think this is how you could add a WASM entrypoint for grammars too: add an alternative loader implementation that just uses an internal list of compiled-in grammars

Reply:
yes that's exactly what I'm attempting now; on the Helix side it looks quite doable, it's all the `tree-sitter-*` libs which will be more tedious to bring to wasm.

Reply:
I got syntax highlighting for `rust`, `toml` and `regex` to work, but it's very hacky:
* hacked `tree-sitter-c2rust` so it uses [`instant`](https://github.com/sebcrozet/instant/) instead of `std::time`
* hacked C code of `tree-sitter-*` crates, for `wasm` compilation + dependency on hacked `tree-sitter-c2rust`; not big changes but still tedious and I dont see a way to automate the process. Maybe `c2rust`ing the language libs, but I have no experience with it.

It's nice to see it working but it's quite a mess, so I'll keep it on the side for now.

Reply:
Why do we need to use `tree-sitter-c2rust` rather than https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/README.md ? I think we should look at how the tree-sitter playground functions since it runs grammars via wasm (and it lazy loads them instead of compiling them into the main binary)

https://tree-sitter.github.io/tree-sitter/playground

Reply:
Oh that's interesting, I hadnt spotted they had the playground in Rust on wasm too.

It's puzzling, as `tree-sitter` actually compiles to `wasm32-unknown-emscripten`, but not to `wasm32-unknown-unknown`. And `wasm32-unknown-emscripten` is not compatible with `wasm-bindgen`: https://rustwasm.github.io/docs/wasm-bindgen/reference/rust-targets.html?highlight=emscripten#other-web-targets.

I'll take a look, it would be so much better this way.

Reply:
Ok so the playground is a Rust web server, serving `emscripten` wasm binaries and regular JS code. The hot loading is js/emscripten code, see https://github.com/tree-sitter/tree-sitter/blob/7c0cee70f55fdfe1667890d372441a85f3127ee6/lib/binding_web/binding.js#L1024-L1042

The Rust bindings compile to linux and wasm-emscripten, but not to wasm-unknown.

I dont see an easy way to interop between a `wasm32-unknown` app and `wasm32-emscripten` libs. Creating Rust bindings for tree-sitter-js and then writing a facade/adapter so Helix code can abstract over the native and web bindings? I got a headache just thinking about it :sweat_smile: 

Open to any ideas!